rudder-core/src/main/scala/com/normation/rudder/domain/queries/CmdbQuery.scala

```diff
@@ -323,6 +325,39 @@ case object MemoryComparator extends CriterionType {
   }
 }
 
+
+case object MachineComparator extends CriterionType {
+
+  val machineTypes = "Virtual machine" ::  "Physical machine" :: Nil
+
+  override def comparators = Seq(Equals, NotEquals)
+  override protected def validateSubCase(v: String, comparator:CriterionComparator) = {
+    if (null == v || v.length == 0) Failure("Empty string not allowed") else Full(v)
+   }
+
+  override def toLDAP(value: String) = Full(value)
+
+  override def buildFilter(attributeName: String, comparator: CriterionComparator, value: String): Filter = {
+    val v = value match {
+      // the machine can't belong to another type
+      case "Virtual machine" => OC_VM
+      case "Physical machine" => OC_PM
+    }
+    comparator match {
+      case Equals => IS(v)
+      case _ => NOT(IS(v))
+    }
+  }
+
+  override def toForm(value: String, func: String => Any, attrs: (String, String)*) : Elem =
+    SHtml.select(
+      (machineTypes map (e => (e, e))).toSeq,
```

Comment:
I know it's not done in the code you used, but, you should write this code like this
<pre>
    SHtml.select(
        (machineTypes map (e => (e, e))).toSeq
      , { if(machineTypes.contains(value)) Full(value) else Empty}
      , func
      , attrs:_*
    )
</pre>